lib/charms/opensearch/v0/opensearch_secrets.py

```diff
@@ -41,21 +41,77 @@
 logger = logging.getLogger(__name__)
 
 
+class SecretUpdateFailedError(Exception):
+    """The exception to be thrown when secret-changed fails to update its secret."""
+
+    pass
+
+
 class OpenSearchSecrets(Object, RelationDataStore):
     """Encapsulating Juju3 secrets handling."""
 
     LABEL_SEPARATOR = ":"
 
-    def __init__(self, charm, peer_relation: str):
+    def __init__(self, charm, peer_relation: str, secrets_to_cb: Dict[str, any] = {}):
         Object.__init__(self, charm, peer_relation)
         RelationDataStore.__init__(self, charm, peer_relation)
 
         self.cached_secrets = SecretCache()
+        self.secrets_to_cb = secrets_to_cb
 
         self.framework.observe(self._charm.on.secret_changed, self._on_secret_changed)
 
+    def _update_secret_value(self, event: SecretChangedEvent):
+        """Updates the secret value.
+
+        This method manages the secret update: compare the old value and
+        the new one (via peek=True). If they are different, then use the callback if
+        available and refreshes the secret. If the callback fails, then it raises an
+        exception to break the hook.
+
+        If the secret has a callback function, it will be called before the secret
+        is finally refreshed.
+        """
+        secret = event.secret
+        cb = self.secrets_to_cb.get(secret.label or secret.id)
+        scope = Scope.APP if Scope.APP in (secret.label or "") else Scope.UNIT
+        if not cb or (scope == Scope.APP and not self._charm.unit.is_leader()):
+            secret.get_content(refresh=True)
+            return
+        # There is a callback function to update the secret
+        old_val = secret.get_content()
+        new_val = secret.peek_content()
+        try:
+            if new_val == old_val:
+                # Nothing to do, these dictionaries have the same content
+                return
+            if cb(old_val, new_val):
+                # Remove the old value
+                secret.get_content(refresh=True)
+            else:
+                # The secret failed to be applied.
+                # Break this hook so it can be retried later on
+                raise SecretUpdateFailedError(
+                    f"Failed to update {secret.label or secret.id} using callback function {cb}"
+                )
+        except (ValueError, model.ModelError) as err:
+            # https://bugs.launchpad.net/juju/+bug/2042596
+            # Only triggered when 'refresh' is set
+            known_model_errors = [
+                "ERROR either URI or label should be used for getting an owned secret but not both",
+                "ERROR secret owner cannot use --refresh",
+            ]
+            if isinstance(err, model.ModelError) and not any(
+                msg in str(err) for msg in known_model_errors
+            ):
+                raise
+            # Due to: ValueError: Secret owner cannot use refresh=True
+            self._secret_content = self.meta.get_content()
+
```

Comment:
Is there a reason to have a new, separate function for the `secret-changed` event specifically, instead of integrating it with the existing `_update_juju_secrets()` function?

Also, I think secrets manipulation should happen via the interface functions of this class, not private methods.

Also, what is this `cb` supposed to refer to...?

I have the gutfeeling that this should rather be merged with the existing functionality rahter than implemented as brand new -- outside of the existing caching method.

Reply:
Hi @juditnovak, as we've discussed, this function is added outside of any other in `opensearch_secrets.py` because I wanted to quickly fix our charm to support 3.1.7 and then, we move to a more permanent solution using our common library.

Callback (or `cb`) is used to distinguish between secrets. With the changes landed in 3.1.7, now some secrets (e.g. passwords) will be able to read both old and new values and execute the change on their database, whereas other secrets (e.g. private keys/certs) can be refreshed right away. The callback is for that distinction:
* private keys / certs: `cb=None`, they just execute the refresh
* passwords: `cb=change_my_password(old_val, new_val)`, they take the longer tour, where we execute the password and, once we are sure it worked, secret-changed proceeds and refresh to the latest secret value